### DISCUSSION TITLE
Question about Message Construction

### DISCUSSION THREAD

**Discussion by lsnty5190** [ACEsuit/mace#102] [Q&A] [answered]
Hi all! Thanks for your work on Multi-ACE. 

The key component of MACE is the construction of A-basis:
<img width="430" alt="image" src="https://user-images.githubusercontent.com/67890032/232690272-6a51dd67-e604-43f9-b8f4-bb7fda158ae2.png">
and B-basis:
<img width="716" alt="image" src="https://user-images.githubusercontent.com/67890032/232690325-5d67a0ff-bab6-4730-be18-900488cb7a71.png">

But I didn't find any **scatter sum** operation for the basis function in the code (the summation operation to construct A-basis), or the **multiplication** of A-basis (to construct the B-basis). The implementation is the direct tensor product between `node_feats` and `edge_feats`, which are augmented by spherical harmonics. 
Am I missing something else? Looking forward to your reply. Thanks!

**Comment by davkovacs** [comment, marked as answer]
Hi! Thanks for your question! 

The A basis is constructed slightly differently in the 1st and 2nd layers, but indeed that is an important component of MACE. The scatter sum operation is part of the interaction block. In the first layer:
https://github.com/ACEsuit/mace/blob/44b6da4a5edaa4c3ef867a11728555403d1d475d/mace/modules/blocks.py#L530

In the subsequent layer(s):
https://github.com/ACEsuit/mace/blob/44b6da4a5edaa4c3ef867a11728555403d1d475d/mace/modules/blocks.py#L603

The B basis is probably best thought of as being part of the update. It is formed on the nodes by taking symmetric tensor products of the A functions. You can find the operations in the product block of MACE, which indeed does not contain a scatter sum:
https://github.com/ACEsuit/mace/blob/44b6da4a5edaa4c3ef867a11728555403d1d475d/mace/modules/blocks.py#L162

Let me know if you have any further questions!

**Reply by lsnty5190** [reply]
Thanks for your answer. So the **A basis** is obtained by the `message aggregation`, and the **B basis** is computed by `SymmetricContraction`. Am I right :) ?

**Reply by davkovacs** [reply]
yes, that sounds right. Note, that the contraction computes the learnable weights times the B basis in one go, for details see the appendix of the paper.

**Comment by lsnty5190** [comment]
Hi davkovacs. Just another brainstorming though. What would happen when two **A basis** from two **adjacent atomic clusters** are multiplied, i.e., $\prod A_i A_j$ ? Would it still represent the many-body expansion or the interaction between two atomic clusters (maybe cluster merging)?

**Reply by ilyes319** [reply]
The product is over A basis on the same nodes, so these products do not occur. The many body expansion is implicitly computed by doing $(\sum_{j \in \mathcal{N}(i)} \phi_{ij})^{\nu}$. So you see, all the interactions are within clusters.

**Reply by davkovacs** [reply]
that would form a different set of many-body functions, that you would need to symmetrise slightly differently. 
Another reason why we don't do that, is because we try to avoid having sum over ij, ie. N^2 scaling. In MACE the expensive operations are always done on the nodes (atoms) not on the edges.

**Reply by lsnty5190** [reply]
Right. I understand the "density trick" in MACE to reduce the computational overhead. But what I think is that maybe $\prod A_i A_j$ as mentioned could implicitly represent the 4-body interaction (still within O(N)), similar to $\prod A_{j1}A_{j2}A_{j3}$?

**Reply by davkovacs** [reply]
right, you get 4-body, but you still need to sum over ij, so not O(N).
